### PR TITLE
Random and Initializer

### PR DESCRIPTION
**Pull request by terrytangyuan**

### CODE REVIEW

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+  def _initBilinear() = {
+    var weight = Array.fill[Float](arr.size)(0.0f)
+    val shape = arr.shape
+    val f = shape(3) / 2.0f
+    val c = (2 * f - 1 - f % 2) / (2.0f * f)
+
+    (0 to (arr.size)).foreach { i =>
+      var x = i % shape(3)
+      var y = (i % shape(3)) % shape(2)
```
Comment (yanqingmen) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: this line would be
    var y = (i/shape(3)) % shape(2)

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+
+/**
+ * Initialize the weight with normal(0, sigma)
+ *
+ * @param name name of corrosponding ndarray
+ * @param arr ndarray to be Initialized
+ * @param sigma Standard deviation for gaussian distribution.
+ */
+class Normal(name: String, arr: NDArray, protected val sigma: Float=0.01f) extends Initializer(name: String, arr: NDArray) {
```
Comment (yanqingmen) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: maybe we do not need to declare the 'name' and 'arr' repeatedly?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    val fanOut = shape(0)
+    var factor = 1
+
+    factor = factorType match {
+      case "avg" => (fanIn + fanOut) / 2
+      case "in" => fanIn
+      case "out" => fanOut
+      case _ => throw new IllegalArgumentException("Incorrect factor type")
+    }
+    val scale = math.sqrt(magnitude / factor) toFloat
```
Comment (yanqingmen) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: missing '.'

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  } else if (name.endsWith("moving_mean")) {
+    _initZero()
+  } else if (name.endsWith("moving_var")) {
+    _initZero()
+  } else if (name.endsWith("moving_avg")) {
+    _initZero()
+  } else {
+    throw new IllegalArgumentException(s"Unknown initialization pattern for ${name}.")
+  }
+
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: __call__ in python should be translated as apply() in scala. What python code does here is to change arr's elements, not its reference.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Random.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+   * @param ctx Context of output NDArray, will use default context if not specified.
+   * @param out Output place holder
+   * @return The result NDArray with generated result.
+   */
+  def uniform(low: Float, high: Float, shape: Array[Int]=null, ctx: Context=null, out: NDArray=null): NDArray = {
+    if (out != null) {
+      require(shape == null & ctx == null, "shape and ctx is not needed when out is specified.")
+    } else {
+      require(shape != null, "shape is required when out is not specified")
+      var out = empty(shape, ctx)
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Random.scala]: this out is in if ... else .... scope, which will be destroyed once it is executed.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Random.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+   * @param ctx Context of output NDArray, will use default context if not specified.
+   * @param out Output place holder
+   * @return The result NDArray with generated result.
+   */
+  def normal(mean: Float, stdvar: Float, shape: Array[Int]=null, ctx: Context=null, out: NDArray=null): NDArray = {
+    if (out != null) {
+      require(shape == null & ctx == null, "shape and ctx is not needed when out is specified.")
+    } else {
+      require(shape != null, "shape is required when out is not specified")
+      var out = empty(shape, ctx)
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Random.scala]: same issue as above

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala (hunk 2 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: I think what @terrytangyuan did here is fine.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  def _initGamma() = {
+    arr = ones(arr.size)
+  }
+
+  def _initBeta() = {
+    arr = zeros(arr.size)
+  }
+
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: arr = ... -> arr.set(...) in these init methods.
change reference here is no use

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ *
+ * Base class for Initializer.
+ *
+ * @author Yuan Tang
+ *
+ * @param name name of corrosponding ndarray
+ * @param arr ndarray to be Initialized
+ */
+abstract class Initializer(protected val name: String, protected var arr: NDArray) {
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: Initializer in python takes no arguments

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala (hunk 2 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: Normal class in python takes sigma as constructor argument only. no name or arr.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Initialize the weight with normal(0, sigma)
+ *
+ * @param name name of corrosponding ndarray
+ * @param arr ndarray to be Initialized
+ * @param sigma Standard deviation for gaussian distribution.
+ */
+class Normal(name: String, arr: NDArray, protected val sigma: Float=0.01f) extends Initializer(name: String, arr: NDArray) {
+  def _initWeight() = {
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: should exists an 'override' here?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    } else if (name.endsWith("moving_var")) {
+      _initZero()
+    } else if (name.endsWith("moving_avg")) {
+      _initZero()
+    } else {
+      throw new IllegalArgumentException(s"Unknown initialization pattern for ${name}.")
+    }
+  }
+
+  def _initBilinear() = {
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/Initializer.scala]: I think always write down the return type for a function is a good habit, even though it's Unit. Besides the clarity, overloaded functions may cause problems when the return type is not specified. This is also a code spec in many scala project, e.g. Apache Spark

**Pull request merged**